### PR TITLE
[WIP] Add Library Manager

### LINKED ISSUE #143: Library Management
Quick proposal, any thoughts?

I'd like libraries not to be shaded directly in the jars. Instead, dependencies could be specified in the glowstone.yml like so:

``` yaml
libraries:
- 'org.xerial:sqlite-jdbc:3.7.2'
- 'mysql:mysql-connector-java:5.1.14' # maybe optional version specifiers
```

These files would be verfied and downloaded if necessary (maybe even from maven central since that already exists, verifying md5+gpg if available) and added to the classpath available to plugins.

At some point this could allow plugins to define their own libraries to be available in the plugin.yml too.

### PR DESCRIPTION
**Pull request by mastercoms**
This adds a library manager. Not much at the moment though, it just downloads libraries.
Will fix #143 
recommended to add `-classpath ./libraries/*` to the start script

todo: 
- [x] download libraries
- [ ] verify libraries
- [ ] load classes without start script modifications
- [x] enable users to choose their own repos
- [ ] download from another repo if the last one does not work

### CODE REVIEW

**Comment by mastercoms** [comment]
There is some extra debug logging stuff that probably shouldn't be pulled.

File: src/main/java/net/glowstone/util/LibraryManager.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import java.nio.channels.ReadableByteChannel;
+import java.util.List;
+import java.util.logging.Level;
+import net.glowstone.GlowServer;
+import static net.glowstone.GlowServer.logger;
+
+
+public final class LibraryManager {
+     /**
+     * The server of this world.
```
Comment (ZephireNZ) [inline src/main/java/net/glowstone/util/LibraryManager.java]: This world?

**Comment by PaulBGD** [comment]
When this is ready, I'd love to use it. One of the most annoying (and bandwidth) using parts of craftbukkit was that you downloaded all of the dependencies.

**Comment by mastercoms** [comment]
@PaulBGD well, it is technically ready right now, I just need to make it automatic.

**Comment by gdude2002** [comment]
Alright, it's building.. Why'd you call it patch-4, out of interest?

EDIT: Built! http://bamboo.gserv.me/browse/GLOW-SRV-PR419-87

**Comment by mastercoms** [comment]
@gdude2002 Github auto-named my branch that.

**Comment by turt2live** [comment]
@mastercoms friendly reminder to limit your scope. We don't want to review large PRs and we don't want you to have to suffer through diffs that take a few hours to read. 

Splitting this out into manageable parts that work mostly independently of each other is fine. From looking at your description, you are taking on too much in a single PR.

Please also remember to collect lots of feedback from the community (and even others outside of it) such that the system is well designed. Just remember to also not bend to everyone's request when they specify a "really cool feature that should be added". 

Thanks!

**Comment by mastercoms** [comment]
@turt2live I'll try not to turn it into Windows Longhorn. I'm open to any reasonable suggestions, but I do understand it cannot be bloated. 
In this PR, should I just include a library downloader, and then include the resolver later in another PR?

**Comment by bendem** [comment]
I'm not really versed into this, but separating the pr in
- lib download
- lib check
- repos
- adding to classpath (loading classes without start script modifications)

might be a good starting point.

Some points can still get merged/splitted depending on the complexity of the task.

**Comment by turt2live** [comment]
I would say that lib checking & downloading (with install on classpath) would be a good start. Then repository stuff in another.

**Comment by mastercoms** [comment]
I've decided to include just a downloader, verifier, and installer in this PR. I'll update it in a few days.
Another PR will include lib resolving.

Also I'm not dead (yay!?)

**Comment by dequis** [comment]
Poke?

**Comment by turt2live** [comment]
@mastercoms psst. This doesn't merge thanks to the commit spam on master.

**Comment by mastercoms** [comment]
@dequis sorry I would love to code this but school has been really painful these days but I'll try to do it over the weekend

**Comment by mastercoms** [comment]
Did I do this conflict resolving correctly? (there seems to be a lot of excess commits)

**Comment by dequis** [comment]
...no.

And I'm not sure how you managed to do that, but all the comment threads from those commits got included in this PR

**Comment by mastercoms** [comment]
@dequis I probably forgot to git reset...
why am I so bad at git

**Comment by dequis** [comment]
Hahah, okay i see what happened here: https://github.com/mastercoms/Glowstone/pull/19

He PR'd a rebase, you merged it like you'd merge a PR, and it became a mess. Kids, don't do this at home.

Anyway, like he said in the comments, you need to `git reset --hard` against it. More precisely:

```
# just to be sure that you're in that branch
git checkout patch-4

git remote add aaron1011 https://github.com/Aaron1011/Glowstone
git fetch aaron1011
git reset --hard aaron1011/patch-4

git push --force
```

**Comment by mastercoms** [comment]
@dequis Ok, I'll try again.
EDIT: it worked

**Comment by dequis** [comment]
\o/

**Comment by mastercoms** [comment]
Actually, something changed in GlowServer.java regarding resource packs /o\

**Comment by dequis** [comment]
That must have been @Aaron1011's manually editing that part to fix style. So, not an issue with the merge. Doesn't belong in this PR but whatever.

**Comment by Aaron1011** [comment]
Sorry about all the problems I caused. I did add a note to the PR saying to `git reset --hard` with my rebased branch, but it wasn't very clear. Next time I'll do something like **DO NOT HIT MERGE** xD

**Comment by dequis** [comment]
Or, you know, like, just use `git merge` and push that and let the other person push the merge button.

That button is actually the main reason I send those merge PRs - often people are semi-AFK, have no access to a commandline, but can use a browser (even from a phone), so pushing the merge button is much easier for them.

**Comment by gdude2002** [comment]
This is failing. http://bamboo.gserv.me/browse/GLOW-SRV-PR419-126

```
18-Nov-2014 07:34:02    Rebasing..
18-Nov-2014 07:34:02    First, rewinding head to replay your work on top of it...
18-Nov-2014 07:34:02    Applying: Corrected formatting on new resource pack methods.
18-Nov-2014 07:34:02    Using index info to reconstruct a base tree...
18-Nov-2014 07:34:02    M        src/main/java/net/glowstone/GlowServer.java
18-Nov-2014 07:34:02    Falling back to patching base and 3-way merge...
18-Nov-2014 07:34:02    Auto-merging src/main/java/net/glowstone/GlowServer.java
18-Nov-2014 07:34:02    CONFLICT (content): Merge conflict in src/main/java/net/glowstone/GlowServer.java
18-Nov-2014 07:34:02    Failed to merge in the changes.
18-Nov-2014 07:34:02    Patch failed at 0001 Corrected formatting on new resource pack methods.
```

**Comment by dequis** [comment]
Yeah...

> That must have been @Aaron1011's manually editing that part to fix style

**Comment by mastercoms** [comment]
@Aaron1011's fix style patch has now been obliterated.
(happy thanksgiving everyone!)

**Comment by mastercoms** [comment]
I was thinking about repurposing the downloader to be separate from the library manager, so I can make threaded downloads. And perhaps, we could use that to self update glowstone, and have plugins use an api to update themselves, not in this PR, but in a future PR. Thoughts?

(and I lost my in progress code for this, so I'll have to rewrite most of it anyway)

**Comment by mastercoms** [comment]
Closing since https://github.com/GlowstoneMC/Glowstone/commit/cb42c013c9ef02ccd003340099e3cf6092c95e74 already fulfills the purpose of this.

**Pull request closed without merging**